contracts/StableJoeStaking.sol

```diff
@@ -0,0 +1,312 @@
+     */
+    function initialize(
+        IERC20Upgradeable _rewardToken,
+        IERC20Upgradeable _joe,
+        address _feeCollector,
+        uint256 _depositFeePercent
+    ) external initializer {
+        __Ownable_init();
+        require(_feeCollector != address(0), "StableJoeStaking: fee collector can't be address 0");
+        require(_depositFeePercent < 5e17, "StableJoeStaking: max deposit fee can't be greater than 50%");
+
+        joe = _joe;
+        depositFeePercent = _depositFeePercent;
+        feeCollector = _feeCollector;
+
+        isRewardToken[_rewardToken] = true;
+        rewardTokens.push(_rewardToken);
+        PRECISION = 1e24;
+    }
+
+    /**
+     * @notice Deposit JOE for reward token allocation
+     * @param _amount The amount of JOE to deposit
+     */
+    function deposit(uint256 _amount) external {
+        UserInfo storage user = userInfo[msg.sender];
+
+        uint256 fee = (_amount * depositFeePercent) / 1e18;
+        uint256 amountMinusFee = _amount.sub(fee);
+
+        uint256 previousAmount = user.amount;
+        uint256 newAmount = user.amount.add(amountMinusFee);
+        user.amount = newAmount;
+
+        uint256 len = rewardTokens.length;
+        uint256 pending;
+        for (uint256 i; i < len; i++) {
+            IERC20Upgradeable token = rewardTokens[i];
+            updateReward(token);
+
+            if (previousAmount > 0) {
+                pending = previousAmount.mul(accRewardPerShare[token]).div(PRECISION).sub(user.rewardDebt[token]);
+                if (pending != 0) safeTokenTransfer(token, msg.sender, pending);
+            }
+            user.rewardDebt[token] = newAmount.mul(accRewardPerShare[token]).div(PRECISION);
+        }
+
+        internalJoeBalance = internalJoeBalance.add(amountMinusFee);
+        joe.safeTransferFrom(msg.sender, feeCollector, fee);
+        joe.safeTransferFrom(msg.sender, address(this), amountMinusFee);
+        emit Deposit(msg.sender, amountMinusFee);
+    }
+
+    /**
+     * @notice get user info
+     * @param _user The address of the user
+     * @param _rewardToken The address of the reward token
+     * @return uint256 the user's amount
+     * @return uint256 the user's reward debt of the reward token chosen
+     */
+    function getUserInfo(address _user, IERC20Upgradeable _rewardToken) external view returns (uint256, uint256) {
+        UserInfo storage user = userInfo[_user];
+        return (user.amount, user.rewardDebt[_rewardToken]);
+    }
+
+    /**
+     * @notice get the number of reward tokens
+     * @return The length of the array
+     */
+    function rewardTokensLength() external view returns (uint256) {
+        return rewardTokens.length;
+    }
+
+    /**
+     * @notice Set the reward token
+     * @param _rewardToken The address of the reward token
+     */
+    function addRewardToken(IERC20Upgradeable _rewardToken) external onlyOwner {
+        require(
```

Comment:
We should include some validation on the maximum amount of reward tokens (this can probably be set quite high) to protect against unbounded list iteration.